src/msg_handler.c

```diff
@@ -0,0 +1,388 @@
+            // Get stored voltage value
+            strncat(response, itoa(battery.voltage, temp_str), max_len);
+            return strlen(response);
+        }
+        strncat(response, "NACK:Unknown battery command", max_len);
+        return strlen(response);
+    } else if (strcmp(next_arg, "BTN") == 0) {
+        next_arg = strtok(NULL, ":");
+        if (next_arg == NULL) {
+            strncat(response, "NACK:Missing button name", max_len);
+            return strlen(response);
+        }
+
+        if (strcmp(next_arg, "START") == 0) {
+            next_arg = strtok(NULL, ":");
+            if (next_arg == NULL) {
+                strncat(response, "NACK:Missing button command", max_len);
+                return strlen(response);
+            }
+
+            if (strcmp(next_arg, "GET?") == 0) {
+                strncat(response, int_button_pressed?"1":"0", max_len);
+                strncat(response, ":", max_len - strlen(response));
+                strncat(response, ext_button_pressed?"1":"0", max_len - strlen(response));
+
+                // Clear button state
+                int_button_pressed = false;
+                ext_button_pressed = false;
+                return strlen(response);
+            }
+
+            strncat(response, "NACK:Invalid button command", max_len);
+            return strlen(response);
+        }
+        strncat(response, "NACK:Invalid button name", max_len);
+        return strlen(response);
+    } else if (strcmp(next_arg, "NOTE") == 0) {
+        next_arg = strtok(NULL, ":");
+        if (next_arg == NULL) {
+            strncat(response, "NACK:Missing note frequency", max_len);
+            return strlen(response);
+        }
+
+        int note_freq;
+
+        if (isdigit((int)next_arg[0])) {
+            note_freq = atoi(next_arg);
+            // bounds check
+            if (note_freq < 0) {
+                strncat(response, "NACK:Invalid note frequency", max_len);
+                return strlen(response);
+            }
+        } else {
+            strncat(response, "NACK:Invalid note frequency", max_len);
+            return strlen(response);
+        }
+
+        next_arg = strtok(NULL, ":");
+        if (next_arg == NULL) {
+            strncat(response, "NACK:Missing note duration", max_len);
+            return strlen(response);
+        }
+
+        int note_dur;
+
+        if (isdigit((int)next_arg[0])) {
+            note_dur = atoi(next_arg);
+            // bounds check
+            if (note_dur < 0) {
+                strncat(response, "NACK:Invalid note duration", max_len);
+                return strlen(response);
+            }
+        } else {
+            strncat(response, "NACK:Invalid note duration", max_len);
+            return strlen(response);
+        }
+
+        // Generate note
+        buzzer_note(note_freq, note_dur);
+
+        strncat(response, "ACK", max_len);
+        return strlen(response);
+    } else if (strcmp(next_arg, "*IDN?") == 0) {
```

Comment:
What's the significance of the asterisk prefix?

Reply:
The protocol is loosely based on the [SCPI](https://en.wikipedia.org/wiki/Standard_Commands_for_Programmable_Instruments) syntax, the asterisk is generally used to denote the common/system commands.